helix-term/src/ui/prompt.rs

```diff
@@ -304,19 +318,29 @@ impl Prompt {
 
         let end = values.len().saturating_sub(1);
 
-        let index = match direction {
-            CompletionDirection::Forward => self.history_pos.map_or(0, |i| i + 1),
-            CompletionDirection::Backward => {
-                self.history_pos.unwrap_or(values.len()).saturating_sub(1)
-            }
-        }
-        .min(end);
+        let search_term = &*self.history_search.get_or_insert_with(|| self.line.clone());
+        let filter = |(_, v): &(usize, &String)| v.starts_with(search_term);
+        let Some((index, _)) = (match direction {
+            CompletionDirection::Forward => values
+                .iter()
+                .enumerate()
+                .skip(self.history_pos.map_or(0, |i| i + 1).min(end))
+                .find(filter),
+            CompletionDirection::Backward => values
+                .iter()
+                .enumerate()
+                .take(self.history_pos.unwrap_or(values.len()))
+                .rev()
+                .find(filter),
+        }) else {
+            return
+        };
```

Comment:
```suggestion
       let filtered = match direction {
            CompletionDirection::Forward => values
                .iter()
                .enumerate()
                .skip(self.history_pos.map_or(0, |i| i + 1).min(end))
                .find(filter),
            CompletionDirection::Backward => values
                .iter()
                .enumerate()
                .take(self.history_pos.unwrap_or(values.len()))
                .rev()
                .find(filter),
        };
        let Some((index, _)) = filtered else {
            return
        };
```

is much clearer IMO, else the `return` is easily missed